### PR TITLE
Implement transfer queues for performing async backfills

### LINKED ISSUE #45461: Add bounded transaction queue for SQL Analysis.
Implement a queue that allows both synchronous and asynchronous delivery of messages. This will be used to schedule SQL analysis where it is missing or stale.

### PR DESCRIPTION
**Pull request by crisptrutski**
Closes https://github.com/metabase/metabase/issues/45461

### Description

This is a fairly generic queue that's come out of discussion around using bounded resources for backfills. It essentially implements a bounded version of Java's built-in `TransactionQueue`, which allows publishers to optionally block until their message has been processed. For non-blocking publishers it behaves like a regular `ArrayBlockingQueue`. 

We may want to rework or abandon this based on the outcome of the related [tech doc](https://www.notion.so/metabase/Minimal-SQL-Analysis-Scheduling-ab8b5a63c1dc4fa3b820173b0cd9fe75) for the SQL analysis backfill.

### CODE REVIEW

File: src/metabase/util/queue.clj [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    (or (locking queued-set
+          (when-let [msg (or (.poll ^Queue async-queue)
+                             (.poll sync-queue block-ms TimeUnit/MILLISECONDS))]
+            (.remove queued-set (:payload msg msg))
+            msg))
+        (do (Thread/sleep ^long sleep-ms)
+            (recur)))))
+
+(defn bounded-transfer-queue
+  "Create a bounded transfer queue, specialized based on the high-level options."
```
Comment (crisptrutski) [inline src/metabase/util/queue.clj:69]: Other options we might provide would be a `merge-fn` for coalescing events (e.g. summing counts) and providing a custom `id` function for deduplicating / merging with.

File: test/metabase/util/queue_test.clj [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+(deftest deduplicating-bounded-blocking-queue-test
+  (doseq [dedupe? [true false]]
+    (let [realtime-event-count 500
+          backfill-event-count 1000
+          capacity             (- realtime-event-count 100)
+          ;; Enqueue background events from oldest to newest
+          backfill-events      (range backfill-event-count)
+          ;; Enqueue realtime events from newest to oldest
+          realtime-events      (take realtime-event-count (reverse backfill-events))
+          queue                (queue/bounded-transfer-queue capacity :sleep-ms 10 :block-ms 10 :dedupe? dedupe?)
```
Comment (piranha) [inline test/metabase/util/queue_test.clj:61]: why naked keys rather than a map though?

**Review by sansolovyov — APPROVED** [review]
I mean, cool! 😁

File: test/metabase/util/queue_test.clj (hunk 2 above)
Comment (crisptrutski) [reply test/metabase/util/queue_test.clj:61]: I forgot which one we prefer 😄

Happy to update the usage, but think I'll leave the overloading in the signature.

Comment (piranha) [reply test/metabase/util/queue_test.clj:61]: Well, I think maps are more honest, they prevent errors (like not supplying value to a key), and also easier to line up nicely 😁

Comment (crisptrutski) [reply test/metabase/util/queue_test.clj:61]: How do they prevent forgetting a key, by adding a schema to the argument?

Comment (piranha) [reply test/metabase/util/queue_test.clj:61]: They will complain about `{:key}` that it lacks value 😁 like, not much of a safe thingie, just naked values don't look as regular.

**Pull request merged**